contrib/redis/go-redis.v9/redis.go

```diff
@@ -0,0 +1,185 @@
+		opt := clientOptions.Options()
+		if opt.Addr == "FailoverClient" {
+			additionalTags = []ddtrace.StartSpanOption{
+				tracer.Tag("out.db", strconv.Itoa(opt.DB)),
+			}
+		} else {
+			host, port, err := net.SplitHostPort(opt.Addr)
+			if err != nil {
+				host = opt.Addr
+				port = "6379"
+			}
+			additionalTags = []ddtrace.StartSpanOption{
+				tracer.Tag(ext.TargetHost, host),
+				tracer.Tag(ext.TargetPort, port),
+				tracer.Tag("out.db", strconv.Itoa(opt.DB)),
+			}
+		}
+	} else if clientOptions, ok := client.(clusterOptions); ok {
+		addrs := []string{}
+		for _, addr := range clientOptions.Options().Addrs {
+			addrs = append(addrs, addr)
+		}
+		additionalTags = []ddtrace.StartSpanOption{
+			tracer.Tag("addrs", strings.Join(addrs, ", ")),
+		}
+	}
+	return additionalTags
+}
+
+func (ddh *datadogHook) DialHook(hook redis.DialHook) redis.DialHook {
+	return func(ctx context.Context, network, addr string) (net.Conn, error) {
+		conn, err := hook(ctx, network, addr)
+		return conn, err
+	}
+}
+
+func (ddh *datadogHook) ProcessHook(hook redis.ProcessHook) redis.ProcessHook {
+	return func(ctx context.Context, cmd redis.Cmder) error {
+		raw := cmd.String()
+		length := strings.Count(raw, " ")
+		p := ddh.params
+		startOpts := make([]ddtrace.StartSpanOption, 0, 7+1+len(ddh.additionalTags)+1) // 7 options below + redis.raw_command + ddh.additionalTags + analyticsRate
+		startOpts = append(startOpts,
+			tracer.SpanType(ext.SpanTypeRedis),
+			tracer.ServiceName(p.config.serviceName),
+			tracer.ResourceName(raw[:strings.IndexByte(raw, ' ')]),
+			tracer.Tag("redis.args_length", strconv.Itoa(length)),
+			tracer.Tag(ext.Component, "redis/go-redis.v9"),
+			tracer.Tag(ext.SpanKind, ext.SpanKindClient),
+			tracer.Tag(ext.DBSystem, ext.DBSystemRedis),
+		)
+		if !p.config.skipRaw {
+			startOpts = append(startOpts, tracer.Tag("redis.raw_command", raw))
+		}
+		startOpts = append(startOpts, ddh.additionalTags...)
+		if !math.IsNaN(p.config.analyticsRate) {
+			startOpts = append(startOpts, tracer.Tag(ext.EventSampleRate, p.config.analyticsRate))
+		}
+		span, ctx := tracer.StartSpanFromContext(ctx, "redis.command", startOpts...)
+
+		err := hook(ctx, cmd)
+
+		var finishOpts []ddtrace.FinishOption
+		if err != nil && err != redis.Nil {
+			finishOpts = append(finishOpts, tracer.WithError(err))
+		}
+		span.Finish(finishOpts...)
+		return err
+	}
+}
+
+func (ddh *datadogHook) ProcessPipelineHook(hook redis.ProcessPipelineHook) redis.ProcessPipelineHook {
+	return func(ctx context.Context, cmds []redis.Cmder) error {
+		raw := commandsToString(cmds)
+		length := strings.Count(raw, " ")
+		p := ddh.params
+		startOpts := make([]ddtrace.StartSpanOption, 0, 8+1+len(ddh.additionalTags)+1) // 8 options below + redis.raw_command + ddh.additionalTags + analyticsRate
+		startOpts = append(startOpts,
+			tracer.SpanType(ext.SpanTypeRedis),
+			tracer.ServiceName(p.config.serviceName),
+			tracer.ResourceName(raw[:strings.IndexByte(raw, ' ')]),
```

Comment:
This could be implemented as `cmds[0].FullName()`, which wouldn't require serializing all the commands.
Though, to be honest, I'm not sure the value of deriving the resource name from the first command, where the pipeline might have an arbitrary number of commands. Why not just use "Pipeline" or "Tx" (if the difference can be detected)?

Reply:
@marcind 
Indeed, as you mentioned, it seems  sufficient to simply use "pipeline" as the resource name. 
I noticed that in the [go-redis OpenTelemetry implementation](https://github.com/redis/go-redis/blob/f30d7c4a25473d8783b8c9c18ac1966c25d98c7e/extra/redisotel/tracing.go#L166), the naming is `"redis.pipeline " +summary`, where the summary consists of up to 10 unique commands. However, retrieving the `summary` requires executing a loop internally using CmdsString(), which may impact performance.

What do you think would be the best approach for us? Is it okay to just use "pipeline" as the resource name?

Reply:
Alternatively, as a compromise, we could consider using "pipeline " followed by the first command's full name, like `"pipeline " + cmds[0].FullName()`

Reply:
I simply modified it to "redis.pipeline". [6dd808a](https://github.com/DataDog/dd-trace-go/pull/1730/commits/6dd808a48fe46fb80ad50d845436a0e53664f553)